projects/notebooks/visualizations/rhods-notebooks-performance/analyze/__init__.py

```diff
@@ -1,26 +1,54 @@
 import json
 import logging
+import numpy as np
 
-import matrix_benchmarking.common as common
+from copy import deepcopy
 
+import matrix_benchmarking.common as common
+import matrix_benchmarking.regression as regression
+import matrix_benchmarking.regression.zscore as zscore
 
 def run():
+
+    logging.info(f"Received {common.LTS_Matrix.count_records()} historic LTS entries")
+    lts_entries = []
+    # Take the first result if they were gathered
+    for entry in common.LTS_Matrix.all_records():
+        if entry.results and type(entry.results) is list:
+            entry.results = entry.results[0]
+            lts_entries.append(entry)
+        else:
+            lts_entries.append(entry)
+
     logging.info(f"Received {common.Matrix.count_records()} new entries")
+
+    number_of_failures = 0
+    settings_to_check = ["version", "repeat"]
```

Comment:
I'm not use to understand exactly what's being checked here, ie how you build the reference data used for the regression checking. We'll need to clarify that, it's important to have that clear

Reply:
also, I don't see clearly where you define the KPI being used for regression. Do you test all of them? 🤔
they won't necessarily have the same direction "of failure", eg a `latency` going higher is a failure, but a `bandwidth` going _lower_ is a failure, so you need to define the regression properties for each of them individually, I guess

Reply:
> I'm not use to understand exactly what's being checked here, ie how you build the reference data used for the regression checking. We'll need to clarify that, it's important to have that clear

I can try to make that clear. Or we can add a note once we finalize it. At the moment it is considering all of the LTS datapoints with the same settings (minus the image version) to be the historical data. The `settings_to_check` is just the settings that we want to check for regression against. We iterate over them to perform the regression testing against each.

> also, I don't see clearly where you define the KPI being used for regression. Do you test all of them? 🤔
they won't necessarily have the same direction "of failure", eg a latency going higher is a failure, but a bandwidth going lower is a failure, so you need to define the regression properties for each of them individually, I guess

Each indicator has an optional parameter called `kpi`. There you can specify a list of KPI's that you'd like to check. I suppose this difference in direction could be implemented in each `RegressionIndicator` we create. However, in the basic `ZScoreIndicator` there is no distinction made. There are two reasons in this simple instance:
1. I figured statistically significant variation in either direction was notable. Having a new image version run significantly faster seems like it would be suspicious? It wouldn't be wrong or bad necessarily but it would be good to make sure we know about them.
2. We also track the direction withing the `RegressionResults`, we can still recover that information if we'd like by filtering out the entries that exhibit regression in the correct direction

Reply:
> Having a new image version run significantly faster seems like it would be suspicious? It wouldn't be wrong or bad necessarily but it would be good to make sure we know about them.

this is linked with the `must remain stable` vs `should remain stable` distinction that I mentioned here and there.
* If `must remain stable`, any positive or negative variation should be flagged. The binary is the same (the hardware also), so the performance shouldn't change significantly.
* If `should_remain_stable`, the binary will change, like version bumps, so the new version can be faster, without need to raise a red flag for it. But it can be a flashy green flag :D